prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -312,63 +313,123 @@ void rxr_pkt_handle_readrsp_recv(struct rxr_ep *ep,
 }
 
 /*  RMA_CONTEXT packet functions */
-void rxr_pkt_handle_rma_context_send_completion(struct rxr_ep *ep,
-						struct rxr_pkt_entry *pkt_entry)
+void rxr_pkt_init_write_context(struct rxr_tx_entry *tx_entry,
+				struct rxr_pkt_entry *pkt_entry)
 {
-	struct rxr_tx_entry *tx_entry = NULL;
-	struct rxr_rx_entry *rx_entry = NULL;
 	struct rxr_rma_context_pkt *rma_context_pkt;
-	int ret;
-
-	assert(rxr_get_base_hdr(pkt_entry->pkt)->version == RXR_PROTOCOL_VERSION);
 
+	pkt_entry->x_entry = (void *)tx_entry;
 	rma_context_pkt = (struct rxr_rma_context_pkt *)pkt_entry->pkt;
+	rma_context_pkt->type = RXR_RMA_CONTEXT_PKT;
+	rma_context_pkt->version = RXR_PROTOCOL_VERSION;
+	rma_context_pkt->context_type = RXR_WRITE_CONTEXT;
+	rma_context_pkt->tx_id = tx_entry->tx_id;
+}
+
+void rxr_pkt_init_read_context(struct rxr_ep *rxr_ep,
+			       struct rxr_rdma_entry *rdma_entry,
+			       size_t seg_size,
+			       struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rma_context_pkt *ctx_pkt;
+
+	pkt_entry->x_entry = rdma_entry;
+	pkt_entry->addr = rdma_entry->addr;
+	pkt_entry->pkt_size = sizeof(struct rxr_rma_context_pkt);
+
+	ctx_pkt = (struct rxr_rma_context_pkt *)pkt_entry->pkt;
+	ctx_pkt->type = RXR_RMA_CONTEXT_PKT;
+	ctx_pkt->flags = 0;
+	ctx_pkt->version = RXR_PROTOCOL_VERSION;
+	ctx_pkt->context_type = RXR_READ_CONTEXT;
+	ctx_pkt->rdma_id = rdma_entry->rdma_id;
+	ctx_pkt->seg_size = seg_size;
+}
 
-	switch (rma_context_pkt->rma_context_type) {
-	case RXR_SHM_RMA_READ:
-	case RXR_SHM_RMA_WRITE:
-		/* Completion of RMA READ/WRTITE operations that apps call */
-		tx_entry = pkt_entry->x_entry;
+static
+void rxr_pkt_handle_rma_read_completion(struct rxr_ep *ep,
+					struct rxr_pkt_entry *context_pkt_entry)
+{
+	struct rxr_tx_entry *tx_entry;
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_rdma_entry *rdma_entry;
+	struct rxr_rma_context_pkt *rma_context_pkt;
+	int inject;
+	ssize_t ret;
 
+	rma_context_pkt = (struct rxr_rma_context_pkt *)context_pkt_entry->pkt;
+	assert(rma_context_pkt->type == RXR_RMA_CONTEXT_PKT);
+	assert(rma_context_pkt->context_type == RXR_READ_CONTEXT);
+
+	rdma_entry = (struct rxr_rdma_entry *)context_pkt_entry->x_entry;
```

Comment:
It is mainly for consistency. Currently all the cq code assume context to be a packet, especially the functions handle errors: rxr_cq_handle_cq_error() rxr_cq_handle_tx_error() and rxr_cq_handle_rx_error. Use rdma_entry as context would cause them to crash.

Reply:
Got it, that seems reasonable.